common/src/main/java/io/netty/util/internal/UTF8Decoder.java

```diff
@@ -0,0 +1,174 @@
+     */
+    /* default */ static final int UTF8_ACCEPT = 0;
+
+    /**
+     * The state the decoder will be in if the decoder detects an invalid character.
+     */
+    /* default */ static final int UTF8_REJECT = 12;
+
+    /**
+     * The state table that maps bytes to character classes.
+     * <p>
+     * <p>
+     * Classes:
+     * <ul>
+     *     <li>00..7f -> 0</li>
+     *     <li>80..8f -> 1</li>
+     *     <li>90..9f -> 9</li>
+     *     <li>a0..bf -> 7</li>
+     *     <li>c0..c1 -> 8</li>
+     *     <li>c2..df -> 2</li>
+     *     <li>e0..e0 -> 10</li>
+     *     <li>e1..ec -> 3</li>
+     *     <li>ed..ed -> 4</li>
+     *     <li>ee..ef -> 3</li>
+     *     <li>f0..f0 -> 11</li>
+     *     <li>f1..f3 -> 6</li>
+     *     <li>f4..f4 -> 5</li>
+     *     <li>f5..ff -> 8</li>
+     * </ul>
+     * </p>
+     */
+    /* default */  static final byte[] CHARACTER_CLASSES = new byte[] {
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9,
+            7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+            8, 8, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+            10, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 3, 3, 11, 6, 6, 6, 5, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8,
+    };
+
+    /**
+     * Maps the state and the character class to the next state
+     */
+    /* default */ static final byte[] TRANSLATION_TABLE = new byte[] {
+            0, 12, 24, 36, 60, 96, 84, 12, 12, 12, 48, 72, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
+            12, 0, 12, 12, 12, 12, 12, 0, 12, 0, 12, 12, 12, 24, 12, 12, 12, 12, 12, 24, 12, 24, 12, 12,
+            12, 12, 12, 12, 12, 12, 12, 24, 12, 12, 12, 12, 12, 24, 12, 12, 12, 12, 12, 12, 12, 24, 12, 12,
+            12, 12, 12, 12, 12, 12, 12, 36, 12, 36, 12, 12, 12, 36, 12, 12, 12, 12, 12, 36, 12, 36, 12, 12,
+            12, 36, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
+    };
+
+    public static String decode(byte[] bytes) {
+        return decode(bytes, 0, bytes.length);
+    }
+
+    public static String decode(byte[] bytes, int offset, int length) {
+        int size = bytes.length;
+        if (MathUtil.isOutOfBounds(offset, length, size)) {
+            throw new IndexOutOfBoundsException(illegalIndex(offset, length, size));
+        }
+        UTF8Processor processor = new UTF8Processor(new char[length]);
+
+        for (int i = 0; i < length; i++) {
```

Comment:
The individual performance techniques you referenced (from [1] I'm assuming?) are fine in isolation but the do not collectively represent what we are discussing. Please run the following benchmark on your machine, and if you see the same results please make the change I suggested.

```
netty/microbench$ mvn clean install -DskipTests=false -Dtest=IterationBenchmark
...
Benchmark                              (size)  Mode  Cnt     Score     Error  Units
IterationBenchmark.indexStartingPoint      10  avgt    5     6.551 ±   0.258  ns/op
IterationBenchmark.indexStartingPoint     100  avgt    5    26.327 ±   0.846  ns/op
IterationBenchmark.indexStartingPoint    1000  avgt    5   266.241 ±   7.867  ns/op
IterationBenchmark.indexStartingPoint   10000  avgt    5  2612.150 ±  80.376  ns/op
IterationBenchmark.zeroStartingPoint       10  avgt    5     6.411 ±   0.045  ns/op
IterationBenchmark.zeroStartingPoint      100  avgt    5    27.447 ±   0.316  ns/op
IterationBenchmark.zeroStartingPoint     1000  avgt    5   310.186 ±   7.896  ns/op
IterationBenchmark.zeroStartingPoint    10000  avgt    5  3042.626 ± 121.119  ns/op
```

```java
@Fork(1)
@Threads(1)
@Measurement(iterations = 5)
@Warmup(iterations = 5)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
public class IterationBenchmark extends AbstractMicrobenchmark {
    @Param({ "10", "100", "1000", "10000" })
    public int size;

    private Random random;
    private byte[] bytes;
    private int offset;
    private int length;

    @Setup(Level.Trial)
    public void setup() {
        random = new Random();
        bytes = new byte[size];
        offset = 0;
        length = bytes.length - offset;
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    public int zeroStartingPoint() {
        int sum = 0;
        for (int i = 0; i < length; ++i) {
            sum += bytes[i + offset];
        }
        return sum;
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    public int indexStartingPoint() {
        int sum = 0;
        for (int i = offset; i < length; ++i) {
            sum += bytes[i];
        }
        return sum;
    }
}
```

```
$java -version
java version "1.8.0_92"
Java(TM) SE Runtime Environment (build 1.8.0_92-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.92-b14, mixed mode)
```

[1] https://wiki.openjdk.java.net/display/HotSpot/PerformanceTechniques